### PR TITLE
WIP Add communications preferences (opt-out) to case detail report

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Add the contact ~communication preferences (opt-out)~ fields to the case details report.

Re-worked to use helper function which adds all contact fields. Also add `Case ID` and make subject optional.

Before
----------------------------------------
Cannot see/filter by contact ~communication preferences~ fields.

After
----------------------------------------
Can see/filter by contact ~communication preferences~ fields.

Technical Details
----------------------------------------
Adds field definitions to report class in standard way using helper function

Comments
----------------------------------------

### CODE REVIEW

**Comment by demeritcowboy** [comment]
It seems like there's a convenience function for this, e.g. https://github.com/civicrm/civicrm-core/blob/52817fca33c632c65e52afba8b0848bc065aebd4/CRM/Report/Form/Contribute/Detail.php#L73-L80

By default it adds a lot of checkboxes, but maybe that's ok?

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
* Developer standards
    * [`r-tech`] __Question__
      * It works as-is, just the question of whether the helper is better or not.
    * [`r-code`] __PASS__
    * [`r-maint`] ____
    * [`r-test`] __PASS__

**Comment by eileenmcnaughton** [comment]
Re the helper - in general I think if there isn't a specific problem it's better to use it - otherwise people don't understand why some fields are in some reports & not others

File: CRM/Report/Form/Case/Detail.php [hunk 1]
```diff
@@ -56,9 +64,12 @@ public function __construct() {
             'no_display' => TRUE,
             'required' => TRUE,
           ],
+          'case_id' => [
```
Comment (eileenmcnaughton) [inline CRM/Report/Form/Case/Detail.php:67]: @mattwire this doesn't have 'type' - it might give e-notices? It's an additional field on the report I think

Comment (mattwire) [reply CRM/Report/Form/Case/Detail.php]: @eileenmcnaughton Thanks - added

**Comment by mattwire** [comment]
@eileenmcnaughton @demeritcowboy Updated to use the helper function which means we add most of the contact fields - that seems like a good thing as it makes the report more universally useful. Also added "Case ID" as that seems pretty important for a case report.

**Comment by eileenmcnaughton** [comment]
This looks good to me - if @demeritcowboy gives this a thumbs up it should be merged

**Comment by demeritcowboy** [comment]
There's a couple things because the alias names have changed. Give me a bit to check a bit more but here's the changes I'm starting with:

```diff
index aa43d143ab..ead34de92b 100644
--- a/CRM/Report/Form/Case/Detail.php
+++ b/CRM/Report/Form/Case/Detail.php
@@ -64,6 +64,9 @@ class CRM_Report_Form_Case_Detail extends CRM_Report_Form {
               'no_display' => TRUE,
               'required' => TRUE,
             ],
+            // Duplicate the field above since it's required for this report to
+            // work, but the user might not want it displayed (although why
+            // wouldn't you?)
             'case_id' => [
               'title' => ts('Case ID'),
               'type' => CRM_Utils_Type::T_INT,
@@ -593,7 +596,7 @@ class CRM_Report_Form_Case_Detail extends CRM_Report_Form {
       if (array_key_exists('civicrm_case_subject', $row)) {
         if ($value = $row['civicrm_case_subject']) {
           $url = CRM_Utils_System::url("civicrm/case/ajax/details",
-            "caseId={$row['civicrm_case_id']}&contactId={$row['civicrm_contact_id']}",
+            "caseId={$row['civicrm_case_id']}&contactId={$row['civicrm_contact_contact_id']}",
             $this->_absoluteUrl
           );
           $rows[$rowNum]['civicrm_case_subject'] = "<a class=\"crm-popup\" href=\"$url\">$value</a>";
@@ -624,15 +627,15 @@ class CRM_Report_Form_Case_Detail extends CRM_Report_Form {
         $rows[$rowNum]['civicrm_activity_last_completed_last_completed_activity_subject'] = ts('(no subject)');
         $entryFound = TRUE;
       }
-      if (array_key_exists('civicrm_contact_client_sort_name', $row) &&
-        array_key_exists('civicrm_contact_id', $row)
+      if (array_key_exists('civicrm_contact_contact_sort_name', $row) &&
+        array_key_exists('civicrm_contact_contact_id', $row)
       ) {
         $url = CRM_Utils_System::url("civicrm/contact/view",
-          'reset=1&cid=' . $row['civicrm_contact_id'],
+          'reset=1&cid=' . $row['civicrm_contact_contact_id'],
           $this->_absoluteUrl
         );
-        $rows[$rowNum]['civicrm_contact_client_sort_name_link'] = $url;
-        $rows[$rowNum]['civicrm_contact_client_sort_name_hover'] = ts("View Contact Summary for this Contact");
+        $rows[$rowNum]['civicrm_contact_contact_sort_name_link'] = $url;
+        $rows[$rowNum]['civicrm_contact_contact_sort_name_hover'] = ts("View Contact Summary for this Contact");
         $entryFound = TRUE;
       }
       if (array_key_exists('civicrm_activity_last_last_activity_activity_type', $row)) {
```

**Comment by demeritcowboy** [comment]
Sorry this is probably getting bigger than you wanted but what also happens is if you have a saved instance, then because the field names have now changed the old instance doesn't include the new field. It doesn't crash, but the output is missing the column.

What you could do, except it maybe partly defeats the purpose of the standard fields, is alter the fields just after setting `_columns` to put the field names back so it's the same as before, e.g.

```php
  $this->_columns['civicrm_contact']['fields']['client_sort_name'] = $this->_columns['civicrm_contact']['fields']['contact_sort_name'];
  unset($this->_columns['civicrm_contact']['fields']['contact_sort_name']);
```

And I admit I didn't check contact_id but I would assume the same for contact_id.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy does the function alterDisplayContactFields help here?

**Comment by demeritcowboy** [comment]
Not for the specific issue here since it doesn't look like it touches those two fields. Either:
* the contact sort_name/id fields need to keep their alias names so that upgrades for saved instances work, or
* alterDisplay needs to be edited to use the new field aliases and you need an upgrade script to update instances, or
* some other creative idea

**Comment by mattwire** [comment]
Closing as the upgrade issue is too complicated. I'll continue using this patch with all clients.

**Pull request closed without merging**